core/src/main/java/org/elasticsearch/common/io/stream/NamedWriteableRegistry.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Registry for {@link NamedWriteable} objects. Allows to register and retrieve prototype instances of writeable objects
+ * given their name.
+ */
+public class NamedWriteableRegistry {
+
+    private final Map<String, InnerRegistry<?>> registry = new HashMap<>();
+
+    /**
+     * Registers a {@link NamedWriteable} prototype given its category
+     */
+    public synchronized <T> void registerPrototype(Class<T> categoryClass, NamedWriteable<? extends T> namedWriteable) {
+        @SuppressWarnings("unchecked")
+        InnerRegistry<T> innerRegistry = (InnerRegistry<T>)registry.get(categoryClass.getName());
```

Comment:
shall we check that the categoryClass is the same as the categoryClass stored in the inner registry? Looking for ways make this even safer